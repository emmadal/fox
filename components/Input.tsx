import React, { useState } from "react";
import {
  TextInput,
  View,
  Text,
  TextInputProps,
  Pressable,
  useColorScheme,
  StyleSheet,
} from "react-native";
import Icon from "@expo/vector-icons/Feather";
import { Colors } from "@/constants/Colors";

interface InputProps extends TextInputProps {
  label?: string;
  placeholder?: string;
  value: string;
  onChangeText: (text: string) => void;
  secureTextEntry?: boolean;
  testID?: string;
  textarea?: boolean;
}

/**
 * Input field component.
 *
 * @param {string} label - The label for the input field.
 * @param {string} placeholder - The placeholder text for the input field.
 * @param {string} value - The current value of the input field.
 * @param {function} onChangeText - The function to call when the input field's value changes.
 * @param {boolean} secureTextEntry - Whether the input field should be secure (password).
 * @param {string} testID - The test ID for the input field.
 * @param {boolean} textarea - Whether the input field is a textarea.
 * @returns {JSX.Element} The input field component.
 */
const Input: React.FC<InputProps> = ({
  label,
  placeholder,
  value,
  onChangeText,
  secureTextEntry = false,
  keyboardType,
  testID,
  textarea,
  ...props
}) => {
  const [showPassword, setShowPassword] = useState(false);
  const colorScheme = useColorScheme();
  const isDarkMode = colorScheme === "dark";

  const togglePasswordVisibility = () => setShowPassword(!showPassword);

  return (
    <View style={styles.container}>
      {label && (
        <Text style={[styles.label, isDarkMode && styles.darkText]}>
          {label}
        </Text>
      )}
      <View
        style={StyleSheet.flatten([
          styles.inputContainer,
          textarea && { height: 97 },
          isDarkMode ? styles.darkContainer : styles.lightContainer,
        ])}
      >
        <TextInput
          style={StyleSheet.flatten([
            styles.input,
            textarea && { textAlignVertical: "top" },
            isDarkMode && styles.darkText,
            secureTextEntry && { paddingRight: 40 },
          ])}
          value={value}
          onChangeText={onChangeText}
          underlineColorAndroid="transparent"
          placeholder={placeholder}
          placeholderTextColor={
            isDarkMode ? Colors.dark.text : Colors.light.tabIconDefault
          }
          keyboardType={keyboardType}
          autoCapitalize="none"
          secureTextEntry={secureTextEntry && !showPassword}
          testID={testID}
          multiline={textarea}
          {...props}
        />
        {secureTextEntry && (
          <Pressable
            onPress={togglePasswordVisibility}
            className="absolute right-5 top-1/2 transform -translate-y-1/2 "
            testID="toggle-password"
          >
            <Icon
              name={showPassword ? "eye-off" : "eye"}
              size={24}
              color={isDarkMode ? Colors.dark.text : Colors.light.text}
            />
          </Pressable>
        )}
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    marginBottom: 12,
  },
  label: {
    fontSize: 16,
    fontWeight: "500",
    marginBottom: 6,
  },
  inputContainer: {
    flexDirection: "row",
    justifyContent: "center",
    borderWidth: 1,
    borderRadius: 8,
    paddingHorizontal: 12,
    height: 57,
  },
  lightContainer: {
    borderColor: "#ccc",
    backgroundColor: "#fff",
  },
  darkContainer: {
    borderColor: "#fff",
    backgroundColor: "#000",
  },
  input: {
    flex: 1,
    fontSize: 16,
    color: "#000",
    textAlign: "auto",
  },
  darkText: {
    color: "#fff",
  },
});

export default Input;
