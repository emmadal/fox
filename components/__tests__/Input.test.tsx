import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import Input from "../Input";

jest.mock("@expo/vector-icons/Feather", () => "Icon");

describe("Input Component", () => {
  const mockOnChangeText = jest.fn();

  it("renders correctly", () => {
    const { toJSON } = render(
      <Input
        placeholder="Enter your username"
        value=""
        onChangeText={mockOnChangeText}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders correctly with a label", () => {
    const { getByText, getByPlaceholderText } = render(
      <Input
        label="Username"
        placeholder="Enter your username"
        value=""
        onChangeText={mockOnChangeText}
      />,
    );

    expect(getByText("Username")).toBeTruthy();
    expect(getByPlaceholderText("Enter your username")).toBeTruthy();
  });

  it("handles text input change", () => {
    const { getByPlaceholderText } = render(
      <Input
        placeholder="Enter your username"
        value=""
        onChangeText={mockOnChangeText}
      />,
    );

    const input = getByPlaceholderText("Enter your username");
    fireEvent.changeText(input, "testuser");
    expect(mockOnChangeText).toHaveBeenCalledWith("testuser");
  });

  it("applies correct keyboard type", () => {
    const { getByTestId } = render(
      <Input
        label="Email"
        placeholder="Enter your email"
        value=""
        onChangeText={mockOnChangeText}
        keyboardType="email-address"
        testID="email-input"
      />,
    );

    const input = getByTestId("email-input");
    expect(input.props.keyboardType).toBe("email-address");
  });

  it("renders secure text input with visibility toggle", () => {
    const { getByPlaceholderText, getByTestId, queryByTestId } = render(
      <Input
        placeholder="Enter your password"
        value=""
        secureTextEntry
        onChangeText={mockOnChangeText}
      />,
    );

    const input = getByPlaceholderText("Enter your password");
    expect(input.props.secureTextEntry).toBe(true);

    // Toggle password visibility
    const toggleButton = getByTestId("toggle-password");
    expect(toggleButton).toBeTruthy();

    fireEvent.press(toggleButton);
    expect(queryByTestId("toggle-password")).toBeTruthy();
    expect(input.props.secureTextEntry).toBe(false);
  });

  it("supports textarea mode", () => {
    const { getByPlaceholderText } = render(
      <Input
        placeholder="Enter your comment"
        value=""
        textarea
        onChangeText={mockOnChangeText}
      />,
    );

    const input = getByPlaceholderText("Enter your comment");
    expect(input.props.multiline).toBe(true);
    expect(input.props.style.textAlignVertical).toBe("top");
  });

  it("applies dark mode styles when in dark mode", () => {
    // Mock useColorScheme to return 'dark'
    jest.mock("react-native/Libraries/Utilities/useColorScheme", () => ({
      __esModule: true,
      default: () => "dark",
    }));

    const { getByTestId } = render(
      <Input
        label="Username"
        placeholder="Enter username"
        value=""
        onChangeText={mockOnChangeText}
        testID="username-input"
      />,
    );
    const input = getByTestId("username-input");
    expect(input.props.style.color).toBe("#fff");
  });
});
