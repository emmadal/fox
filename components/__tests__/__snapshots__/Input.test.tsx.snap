// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component renders correctly 1`] = `
<View
  style={
    {
      "marginBottom": 12,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#fff",
        "borderColor": "#ccc",
        "borderRadius": 8,
        "borderWidth": 1,
        "flexDirection": "row",
        "height": 57,
        "justifyContent": "center",
        "paddingHorizontal": 12,
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      onChangeText={[MockFunction]}
      placeholder="Enter your username"
      placeholderTextColor="#687076"
      secureTextEntry={false}
      style={
        {
          "color": "#000",
          "flex": 1,
          "fontSize": 16,
          "textAlign": "auto",
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
</View>
`;
